fendermint/vm/topdown/src/vote/mod.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod operation;
+
+use crate::sync::TopDownSyncEvent;
+use crate::vote::operation::{OperationMetrics, OperationStateMachine};
+use crate::BlockHeight;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::sync::{broadcast, mpsc};
+
+#[derive(Clone)]
+pub struct VoteRecord {}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Config {
+    /// The reactor request channel buffer size
+    req_channel_buffer_size: usize,
+    /// The number of requests the reactor should process per run before handling other tasks
+    req_batch_processing_size: usize,
+    /// The number of vote recording requests the reactor should process per run before handling other tasks
+    gossip_req_processing_size: usize,
+    /// The time to sleep for voting loop if nothing happens
+    voting_sleep_interval_sec: u64,
+}
+
+pub struct VoteReactorClient {
+    tx: mpsc::Sender<VoteReactorRequest>,
+}
+
+pub fn start_vote_reactor(
+    config: Config,
+    gossip_rx: broadcast::Receiver<VoteRecord>,
+    gossip_tx: mpsc::Sender<VoteRecord>,
+    internal_event_listener: broadcast::Receiver<TopDownSyncEvent>,
+) -> VoteReactorClient {
+    let (tx, rx) = mpsc::channel(config.req_channel_buffer_size);
+
+    tokio::spawn(async move {
+        let sleep = Duration::new(config.voting_sleep_interval_sec, 0);
+
+        let inner = VotingHandler {
+            req_rx: rx,
+            gossip_rx,
+            gossip_tx,
+            internal_event_listener,
+            config,
+        };
+        let mut machine = OperationStateMachine::new(inner);
+        loop {
+            machine = machine.step();
+            tokio::time::sleep(sleep).await;
+        }
+    });
+
+    VoteReactorClient { tx }
+}
+
+enum VoteReactorRequest {
+    QueryOperationMode,
+    QueryVotes(BlockHeight),
+}
+
+struct VotingHandler {
+    /// Handles the requests targeting the vote reactor, could be querying the
+    /// vote tally status and etc.
+    req_rx: mpsc::Receiver<VoteReactorRequest>,
+    /// Receiver from gossip pub/sub, mostly listening to incoming votes
+    gossip_rx: broadcast::Receiver<VoteRecord>,
+    gossip_tx: mpsc::Sender<VoteRecord>,
+    /// Listens to internal events and handles the events accordingly
+    internal_event_listener: broadcast::Receiver<TopDownSyncEvent>,
+    config: Config,
+}
+
+impl VotingHandler {
+    fn handle_request(&self, _req: VoteReactorRequest) {}
+
+    fn record_vote(&self, _vote: VoteRecord) {}
+
+    fn handle_event(&self, _event: TopDownSyncEvent) {}
+
+    fn process_external_request(&mut self, _metrics: &OperationMetrics) -> usize {
+        let mut n = 0;
+        while n < self.config.req_batch_processing_size {
+            match self.req_rx.try_recv() {
```

Comment:
I'm not sure I understand the purpose of doing `try_recv()`. If the channel is empty, it will simply proceed to a fixed-duration sleep, which seems inefficient.

Reply:
Initially I'm thinking of using `select!`, but then the drawback I think would be unable to decide execution order. Especially when different channels need to be processed simultaneously. Another way I'm thinking of is directly implement `Future` to poll ourselves. But this way is similar to handling poll.

Reply:
It's possible to reduce the sleep time to make the processing more responsive. In a way, the reactor loop should be priority based. I think `select!` is fair. I have encountered cases where more important messages are delayed.

Reply:
This comment spans also https://github.com/consensus-shipyard/ipc/pull/1136#discussion_r1882055401 & https://github.com/consensus-shipyard/ipc/pull/1136#discussion_r1882029783.

`select!` would indeed imply random order/priority between channels processing. 
That's why in https://github.com/consensus-shipyard/ipc/pull/1136#discussion_r1882055401 I asked about the need for sequential steps: `req_rx.try_recv()`, then `gossip_rx.try_recv()`, then `internal_event_listener.try_recv()`. 

Unless there's a good reason for that, it just creates a more complicated design.

Reply:
Can you elaborate more on that? 

You don't really get this guarantee with the current design, only in the scope of a single `advance()` iteration, for a given batch processing size config (`req_channel_buffer_size`, `req_batch_processing_size`).

It makes sense to employ a design like this only if there's a timeline dependency between the state machine handling of different channels, which is a fragile design, and I don't think is the case. 

If the channels output timing assumes to be non-deterministic, it can similarly be handled by separate actors, or using `select!`.
If they are expected to be timed/ordered, please elaborate more on that.

Reply:
* i’m not sure about the need for `step()`, `advance()`, etc. it seems to be used to control the Paused/Active modes transition and to enable/disable certain operations. but i think it can be simplified, using some observability on that state transition. 
* i assume channels output timing to be arbitrary, and so the ordering of the processing can be considered random as well. access serialization, either with channels, or alternatively with mutex, is only to prevent data races (or, to get your code to compile). what i mean is that when you query a proposal, there’s no logical conditions on including a certain `add_vote` either before or after. that’s why i don’t think that trying to control the channels processing order, or introducing batch processing size, helps. 
* processing multiple channels with `select!` or overloading them onto one with multiple producers is roughly equivalent, because eventually all write events must be serialized to be processed.
* a potential optimization is to separate reads and writes, and allow reads parallelization instead of having it be serialized with the writes. but i'm not sure it makes sense, considering the read/write ratio.
* if we're concerned about the interpreter proposal query latency, we can employ a shared priority channel, and a separate worker to source it from the multiple channels.

Reply:
@LePremierHomme I have updated accordingly. This PR only switches to `select!`, following one will update the rest to ensure a stronger guarantee.

Reply:
Few remarks: 
* there should be some cancellation mechanism, at least for a graceful shutdown, even if currently unused.
* i’m still not sure about `OperationStateMachine` as the base state using `step()`/`advance()` to transition, when transitioning is only to disable/enable a subset of the handlings. perhaps this can be structured differently by spawning an additional actor which observes that state.

Reply:
For the first point, I think I can add them in the later PRs, since the structure is setup.
For the second one, I'm not so sure it's actually a problem at the current stage. Because each mode at the current stage is just a slightly different. Given more operation mode, it might be making sense to break it into different struct. But I feel we can do that pretty easily when the use cases are required.